process/publisher.go

```diff
@@ -0,0 +1,137 @@
+package process
+
+import (
+	"context"
+
+	"github.com/multiversx/mx-chain-core-go/core/check"
+	"github.com/multiversx/mx-chain-notifier-go/data"
+)
+
+type publisher struct {
+	handler PublisherHandler
+
+	broadcast                     chan data.BlockEvents
+	broadcastRevert               chan data.RevertBlock
+	broadcastFinalized            chan data.FinalizedBlock
+	broadcastTxs                  chan data.BlockTxs
+	broadcastBlockEventsWithOrder chan data.BlockEventsWithOrder
+	broadcastScrs                 chan data.BlockScrs
+
+	cancelFunc func()
+	closeChan  chan struct{}
+}
+
+// NewPublisher will create a new publisher component
+func NewPublisher(handler PublisherHandler) (*publisher, error) {
+	if check.IfNil(handler) {
+		return nil, ErrNilPublisherHandler
+	}
+
+	p := &publisher{
+		handler:                       handler,
+		broadcast:                     make(chan data.BlockEvents),
+		broadcastRevert:               make(chan data.RevertBlock),
+		broadcastFinalized:            make(chan data.FinalizedBlock),
+		broadcastTxs:                  make(chan data.BlockTxs),
+		broadcastScrs:                 make(chan data.BlockScrs),
+		broadcastBlockEventsWithOrder: make(chan data.BlockEventsWithOrder),
+		closeChan:                     make(chan struct{}),
+	}
+
+	return p, nil
+}
+
+// Run is launched as a goroutine and listens for events on the exposed channels
+func (p *publisher) Run() {
```

Comment:
I think we need better protection on this function so we can not call it multiple times. Otherwise, it will lead to resource leaks.
Alternatively, we have this go-routine-wrapper component https://github.com/multiversx/mx-sdk-go/blob/main/core/polling/pollingHandler.go maybe you can take the best practices from there.

Reply:
added additional protection